dotnet/Razorvine.Pyrolite/Pyrolite/Pickle/Pickler.cs

```diff
@@ -1,773 +1,128 @@
-				}
-				else
-				{
-					outs.WriteByte(Opcodes.LONG_BINGET);
-                    BinaryPrimitives.WriteInt32LittleEndian(byteBuffer, memo_index);
-					outs.Write(byteBuffer, 0, sizeof(int));
-				}
-				return true;
-			}
-		}
-		return false;
-	}
-
-	/**
-	 * Process a single object to be pickled.
-	 */
-	private bool dispatch(Type t, object o) {
-        Debug.Assert(t != null);
-        Debug.Assert(o != null);
-        Debug.Assert(t == o.GetType());
-
-        // is it a primitive array?
-        if (o is Array) {
-			Type componentType=t.GetElementType();
-			if(componentType != null && componentType.IsPrimitive) {
-				put_arrayOfPrimitives(componentType, o);
-			} else {
-				put_arrayOfObjects((object[])o);
-			}
-			return true;
-		}
-
-        // first check for enums, as GetTypeCode will return the underlying type.
-		if(o is Enum) {
-			put_string(o.ToString());
-			return true;
-		}
-		
-		// first the primitive types
-        switch (Type.GetTypeCode(t))
+        public const int HIGHEST_PROTOCOL = 2;
+        public const int MAX_RECURSE_DEPTH = 200;
+        public const int PROTOCOL = 2;
+
+        protected static readonly IDictionary<Type, IObjectPickler> customPicklers = new Dictionary<Type, IObjectPickler>();
+        protected readonly bool useMemo = true;
+        private IPicklerImplementation _picklerImplementation;
+
+        /// <summary>
+        /// Create a Pickler.
+        /// </summary>
+        public Pickler() : this(true) { }
+
+        /// <summary>
+        /// Create a Pickler. 
+        /// </summary>
+        /// <param name="useMemo">Specify if it is to use a memo table or not.</param>
+        /// <remarks>
+        /// The memo table is NOT reused across different calls.
+        /// If you use a memo table, you can only pickle objects that are hashable.
+        /// </remarks>
+        public Pickler(bool useMemo) => this.useMemo = useMemo;
+
+        /// <summary>
+        /// Close the pickler stream, discard any internal buffers.
+        /// </summary>
+        public void Dispose() => close();
+
+        /// <summary>
+        /// Close the pickler stream, discard any internal buffers.
+        /// </summary>
+        public void close() => _picklerImplementation.Dispose();
+
+        /// <summary>
+        /// Register additional object picklers for custom classes.
+        /// If you register an interface or abstract base class, it means the pickler is used for 
+        /// the whole inheritance tree of all classes ultimately implementing that interface or abstract base class.
+        /// If you register a normal concrete class, the pickler is only used for objects of exactly that particular class.
+        /// </summary>
+        /// <param name="class">the custom class</param>
+        /// <param name="pickler">additional object picklers</param>
+        public static void registerCustomPickler(Type @class, IObjectPickler pickler) => customPicklers[@class] = pickler;
```

Comment:
Just calling out this potential source breaking change `clazz` to `@class`. If someone was calling this method with a named parameter.

Reply:
I want it to remain ``clazz`` for this reason (despite the awkward spelling)

Reply:
Good point, I have reversed this change.